### PR TITLE
[StandardToHandshake] Prototypical feedforward guard implementation

### PR DESCRIPTION
**Pull request by Dinistro**
This PR aims to introduce the guard mechanism for the feed-forward task pipelining transformation.

Related to https://github.com/llvm/circt/issues/3429

This PR is for now a draft to ensure that @mortbopet and I can discuss certain aspects of it. Note that I did not yet change all the affected tests, as the exact details of the transformation aren't fixed yet.

### CODE REVIEW

File: lib/Conversion/StandardToHandshake/StandardToHandshake.cpp [hunk 1]
```diff
@@ -1055,6 +1185,9 @@ static LogicalResult checkUseCount(Operation *op, Value res) {
 
 // Checks if op successors are in appropriate blocks
 static LogicalResult checkSuccessorBlocks(Operation *op, Value res) {
```
Comment (Dinistro) [inline lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:1187]: I'm not sure how we should handle cases like the one involving the buffer. The buffer isn't part of any basic block, thus this check breaks down.

Comment (Dinistro) [reply lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:1187]: We could drop this in favor of the tests.

File: lib/Conversion/StandardToHandshake/StandardToHandshake.cpp [hunk 2]
```diff
@@ -614,6 +614,136 @@ static int getBranchCount(Value val, Block *block) {
...
+  Value newCtrl = rewriter.create<handshake::MuxOp>(ctrlMerge->getLoc(), buf,
+                                                    ctrlMerge.getOperands());
+
+  hl.setBlockEntryControl(mergeBlock, newCtrl);
+
+  // TODO why does the rewriter not figure this out?
+  // Replace with new ctrl value from mux and the index
+  // rewriter.replaceOp(ctrlMerge, {newCtrl, buf});
+  ctrlMerge.result().replaceAllUsesWith(newCtrl);
+  ctrlMerge.index().replaceAllUsesWith(buf);
```
Comment (Dinistro) [inline lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:743]: Ensure that we have correct types.

**Comment by Dinistro** [comment]
This should now contain all the fixes and updated tests. 
I'll have to think about how we can cheaply detect irreducible control flow to gracefully fail in these cases.

File: lib/Conversion/StandardToHandshake/StandardToHandshake.cpp [hunk 3]
```diff
@@ -614,6 +615,191 @@ static int getBranchCount(Value val, Block *block) {
...
+  Value condAsIndex = rewriter.create<arith::IndexCastOp>(
+      cond.getLoc(), rewriter.getIndexType(), cond);
+
+  // TODO how to size these?
+  // Longest path in a CFG-DAG would be O(#blocks)
+  auto buffer = rewriter.create<handshake::BufferOp>(
+      cond.getLoc(), rewriter.getIndexType(), /*size=*/2, condAsIndex,
+      /*bufferType=*/BufferTypeEnum::fifo);
+
+  return rewriter.create<handshake::BranchOp>(cond.getLoc(), buffer);
```
Comment (Dinistro) [inline lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:779]: Does this branch even do anything meaningful? The buffer not being part of a CFG "Block", a branching operation does not seem to have any meaning.

Comment (mortbopet) [reply lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:794]: Handshake branch ops should in general have some thought poured into them... As far as I know, they're left over from the Dynamatic reimplementation. Branch ops don't do anything, and are always removed through canonicalization! https://github.com/llvm/circt/blob/6515ddcb70e11fca5f1e277a95ee263cf2391712/include/circt/Dialect/Handshake/HandshakeCanonicalization.td#L25-L26

File: lib/Conversion/StandardToHandshake/StandardToHandshake.cpp [hunk 4]
```diff
@@ -614,6 +615,206 @@ static int getBranchCount(Value val, Block *block) {
...
+    for (auto pred : mergeBlock->getPredecessors()) {
+      if (info && info->inLoop.contains(pred))
+        continue;
+      nonLoopPreds++;
+    }
+    if (nonLoopPreds > 2)
+      return parentOp
+                 ->emitError("expected a merge block to have two predecessors. "
+                             "Did you run the "
+                             "merge block insertion pass?")
```
Comment (mortbopet) [inline lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:743]: formatting

**Review by mortbopet — APPROVED** [review]

File: lib/Conversion/StandardToHandshake/StandardToHandshake.cpp (hunk 3 above)
Comment (Dinistro) [reply lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:791]: I saw that they are lowered to identity circuits, yeah. 
Then we do not require one here, I guess.

**Pull request merged**